### ISSUE TITLE
TomatoCart v14.0 LXC guest (on TKL LXC app) fails with 500 error

### ISSUE THREAD

**Issue by Dude4Linux** [turnkeylinux/tracker#562]
My recent testing with Ansible and LXC shows consistent failures of the v14.0 TomatoCart image. The web browser indicates an http 500 error was received and the apache error.log shows:

    [Mon Feb 01 19:29:39.353439 2016] [:error] [pid 690] [client 192.168.121.1:57892] PHP Warning:  include(includes/modules/services/MODULE_SERVICES_INSTALLED.php): failed to open stream: No such file or directory in /var/www/tomatocart/includes/classes/services.php on line 62
    [Mon Feb 01 19:29:39.353531 2016] [:error] [pid 690] [client 192.168.121.1:57892] PHP Warning:  include(): Failed opening 'includes/modules/services/MODULE_SERVICES_INSTALLED.php' for inclusion (include_path='.:/usr/share/php:/usr/share/pear') in /var/www/tomatocart/includes/classes/services.php on line 62
    [Mon Feb 01 19:29:39.353610 2016] [:error] [pid 690] [client 192.168.121.1:57892] PHP Fatal error:  Call to a member function load() on null in /var/www/tomatocart/index.php on line 18

dmesg returns one warning that might be related to the lxc container configuration:

     cgroup: new mount options do not match the existing superblock, will be ignored

I haven't had a chance to test an iso install to see if the problem appears there as well.

**Comment by JedMeister** [comment]
Thanks for reporting John. However, it's weird. I was pretty sure that I had tested it (manually) previously and that it had worked ok.

To be sure I downloaded the template and fired it up on my Proxmox server. It all works fine for me...?!

Here is my Apache logs after starting:
````
# cat /var/log/apache2/error.log 
[Fri Sep 04 05:04:52.984171 2015] [mpm_prefork:notice] [pid 29109] AH00163: Apache/2.4.10 (Debian) OpenSSL/1.0.1k configured -- resuming normal operations
[Fri Sep 04 05:04:52.988437 2015] [core:notice] [pid 29109] AH00094: Command line: '/usr/sbin/apache2'
[Fri Sep 04 05:05:06.725089 2015] [mpm_prefork:notice] [pid 29109] AH00169: caught SIGTERM, shutting down
[Tue Feb 02 22:30:02.072624 2016] [mpm_prefork:notice] [pid 2863] AH00163: Apache/2.4.10 (Debian) OpenSSL/1.0.1k configured -- resuming normal operations
[Tue Feb 02 22:30:02.072823 2016] [core:notice] [pid 2863] AH00094: Command line: '/usr/sbin/apache2'
[Tue Feb 02 22:34:56.200178 2016] [mpm_prefork:notice] [pid 2863] AH00169: caught SIGTERM, shutting down
[Tue Feb 02 22:34:57.754629 2016] [mpm_prefork:notice] [pid 11896] AH00163: Apache/2.4.10 (Debian) OpenSSL/1.0.1k configured -- resuming normal operations
[Tue Feb 02 22:34:57.754722 2016] [core:notice] [pid 11896] AH00094: Command line: '/usr/sbin/apache2'
````

I also tested Admin login (https://<ip-or-fqdn>/admin) and that worked fine too...

I wonder what is causing these issues when you are testing?

**Comment by Dude4Linux** [comment]
@JedMeister Thanks for checking.  If you're not seeing the problem on Proxmox 4, then it must be an issue with the LXC container config.  I've been expecting that we will need to provide some custom handling for certain appliances.  Certainly, the LXC image will need special handling to run in nested mode.  I think the best way to handle customization is to follow LXC's lead and use files named something like

    /usr/share/lxc/config/turnkey.<appliance>.<version>.conf

that are included by the lxc-turnkey template whenever the appliance and versions match.

When I get a chance, I'll take a deeper look at TomatoCart to see what changes are needed.  My understanding is that Anton @qrntz is looking into GitLab and several other appliances, so I haven't done any more work on those.  He should be aware of this issue in case there are some common factors that are affecting the other appliances.

**Comment by JedMeister** [comment]
Thanks John. Anton has tidied up GitLab and some others so I think all the major issues are fixed. As you say though we may need to provide some additional config for some specific appliances.

**Comment by Dude4Linux** [comment]
The cgroup: error is not unique to the TomatoCart appliance.  I checked several other appliances and found it was occurring there as well.  I tried several suggestions regarding the use of cgroups within containers, but none were effective in eliminating this error.  According to this bug report, https://bugs.launchpad.net/ubuntu/+source/libcgroup/+bug/1455358, the problem is likely caused by an upstream kernel bug.
The investigation revealed some questions about the proper use of cgroups within containers.  Because they are not specific to the TomatoCart appliance, I'll open a separate issue to carry on that discussion.  The remainder of this thread should focus on the cause of the 500 error on both web ports.

**Comment by Dude4Linux** [comment]
In my testing, I was mistakenly led to believe that the changes I made to handle cgroups within containers fixed the issue with TomatoCart.  The regression tests I ran last night showed that this was not the case, but they gave me a clue about where to look.  The Ansible playbook I use for testing has three parts, create a container, test the container, and destroy the container.  In normal testing, the three parts are run in sequence with no delay between parts.  I used to have programmed delays but they were removed to speed up the testing.
Yesterday, I made changes to the way cgroups are handled, then ran just the create part to check the results manually.  Surprisingly, I found that although the dmesg cgroup error was still present, the TomatoCart was working perfectly.  Last night's regression test and more this morning show that when run without delay, the appliance locks up and returns 500 errors, but if a delay of a minute or more is allowed between the create and the test, the appliance works normally.  My conclusion is that attempting to access the web ports (80 or 443) during the initialization phase causes the appliance to lockup.

Because this is a problem with the test procedure and not the TomatoCart appliance, I am closing this ticket.  The resolution will be to add a programmed delay for certain appliances that exhibit this problem.
